File: solutions/cpp/buy_and_sell_stock_twice.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <array>
#include <cassert>
#include <iostream>
#include <limits>
#include <random>
#include <vector>

using std::array;
using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::min;
using std::numeric_limits;
using std::random_device;
using std::uniform_int_distribution;
using std::uniform_real_distribution;
using std::vector;

// @include
double BuyAndSellStockTwice(const vector<double>& prices) {
  double max_total_profit = 0;
  vector<double> first_buy_sell_profits(prices.size(), 0);
  double min_price_so_far = numeric_limits<double>::max();

  // Forward phase. For each day, we record maximum profit if we
  // sell on that day.
  for (int i = 0; i < prices.size(); ++i) {
    min_price_so_far = min(min_price_so_far, prices[i]);
    max_total_profit = max(max_total_profit, prices[i] - min_price_so_far);
    first_buy_sell_profits[i] = max_total_profit;
  }

  // Backward phase. For each day, find the maximum profit if we make
  // the second buy on that day.
  double max_price_so_far = numeric_limits<double>::min();
  for (int i = prices.size() - 1; i > 0; --i) {
    max_price_so_far = max(max_price_so_far, prices[i]);
    max_total_profit =
        max(max_total_profit,
            max_price_so_far - prices[i] + first_buy_sell_profits[i - 1]);
  }
  return max_total_profit;
}
// @exclude

double BuyAndSellStockTwiceConstantSpace(const vector<double>& prices) {
  array<double, 2> min_prices = {numeric_limits<double>::max(),
                                 numeric_limits<double>::max()},
                   max_profits = {0.0, 0.0};
  for (const double& price : prices) {
    for (int i = 1; i >= 0; --i) {
      max_profits[i] = max(max_profits[i], price - min_prices[i]);
      min_prices[i] =
          min(min_prices[i], price - (i - 1 >= 0 ? max_profits[i - 1] : 0.0));
    }
  }
  return max_profits[1];
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 100);
      n = dis(gen);
    }
    vector<double> A;
    uniform_real_distribution<double> dis(0.0, 10000.0);
    for (int i = 0; i < n; ++i) {
      A.emplace_back(dis(gen));
    }
    assert(fabs(BuyAndSellStockTwiceConstantSpace(A) -
                BuyAndSellStockTwice(A)) < 1.0e-8);
  }
  return 0;
}

File: static/java/BuyAndSellStockTwice.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class BuyAndSellStockTwice {
  // @include
  public static double buyAndSellStockTwice(List<Double> prices) {
    double maxTotalProfit = 0.0;
    List<Double> firstBuySellProfits = new ArrayList<>();
    double minPriceSoFar = Double.MAX_VALUE;

    // Forward phase. For each day, we record maximum profit if we
    // sell on that day.
    for (int i = 0; i < prices.size(); ++i) {
      minPriceSoFar = Math.min(minPriceSoFar, prices.get(i));
      maxTotalProfit = Math.max(maxTotalProfit, prices.get(i) - minPriceSoFar);
      firstBuySellProfits.add(maxTotalProfit);
    }

    // Backward phase. For each day, find the maximum profit if we make
    // the second buy on that day.
    double maxPriceSoFar = Double.MIN_VALUE;
    for (int i = prices.size() - 1; i > 0; --i) {
      maxPriceSoFar = Math.max(maxPriceSoFar, prices.get(i));
      maxTotalProfit = Math.max(
          maxTotalProfit,
          maxPriceSoFar - prices.get(i) + firstBuySellProfits.get(i - 1));
    }
    return maxTotalProfit;
  }
  // @exclude

  private static double buyAndSellStockTwiceConstantSpace(List<Double> prices) {
    List<Double> minPrices = Arrays.asList(Double.MAX_VALUE, Double.MAX_VALUE),
                 maxProfits = Arrays.asList(0.0, 0.0);
    for (Double price : prices) {
      for (int i = 1; i >= 0; --i) {
        maxProfits.set(i,
                       Math.max(maxProfits.get(i), price - minPrices.get(i)));
        minPrices.set(
            i,
            Math.min(minPrices.get(i),
                     price - (i - 1 >= 0 ? maxProfits.get(i - 1) : 0.0)));
      }
    }
    return maxProfits.get(1);
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = gen.nextInt(100) + 1;
      }
      List<Double> A = new ArrayList<>();
      for (int i = 0; i < n; ++i) {
        A.add(gen.nextDouble() * 10000);
      }
      assert(Double.compare(buyAndSellStockTwiceConstantSpace(A),
                            buyAndSellStockTwice(A))
                 == 0
             || Math.abs(buyAndSellStockTwiceConstantSpace(A)
                         - buyAndSellStockTwice(A))
                    < 1.0e-8);
    }
  }
}
